# HG changeset patch
# User Ken Brown <kbrown@cornell.edu>
# Date 1426182319 14400
#      Thu Mar 12 13:45:19 2015 -0400
# Node ID 9ff524a1f78ab7fcf5cb25ca55efbf27c0adac61
# Parent  32d6c4cfcef539fd96cfdd7971050d3848f31048
Make dynamic loading work on Cygwin

* src/lispbibl.d [DYNAMIC_MODULES]: Treat Cygwin the same as non-win32
systems.
* src/clisp-link.in:
* src/makemake.in: Fix dynamic loading on Cygwin.

diff -r e306f0da362f src/clisp-link.in
--- a/src/clisp-link.in Fri Mar 13 07:57:45 2015 -0400
+++ b/src/clisp-link.in Fri Mar 13 11:50:24 2015 -0400
@@ -597,6 +597,9 @@
       # in the build directory, avoid "cp: `...' and `...' are the same file"
       if [ "${absolute_moduledir}" != "${absdestdir}/$moduledir" ]; then
         make clisp-module-distrib LN=${INSTALL_DATA-cp} distribdir="${absdestdir}/$moduledir"
+       if [ "${HSYSOS}" = cygwin ] ; then
+          ${INSTALL_DATA-cp} *.dll.a ${absdestdir}/${moduledir}
+       fi
         mkdir -p ${absdestdir}/${DYNMOD};
         for f in ${DYNDIR_FILES}; do
           cp ${f} ${absdestdir}/${DYNMOD}/
diff -r 32d6c4cfcef5 -r 9ff524a1f78a src/lispbibl.d
--- a/src/lispbibl.d	Thu Mar 12 13:23:02 2015 -0400
+++ b/src/lispbibl.d	Thu Mar 12 13:45:19 2015 -0400
@@ -559,7 +559,7 @@
    Info: resolving _mv_space by linking to __imp__mv_space (auto-import)
    on woe32. */
 #if defined(DYNAMIC_MODULES)
-  #if defined(WIN32_NATIVE) || defined(UNIX_CYGWIN32)
+  #if defined(WIN32_NATIVE)
     #define DYNAMIC_TABLES 1
     #define modexp  __declspec(dllexport)
     #define modimp  __declspec(dllimport)
@@ -568,6 +568,9 @@
     #define DYNAMIC_TABLES 0
     #define modexp
     #define modimp  extern
+    #if defined(UNIX_CYGWIN32)
+      #define EXECUTABLE_NAME "lisp.exe"
+    #endif
   #endif
 #else
   #define DYNAMIC_TABLES 0
diff -r 32d6c4cfcef5 -r 9ff524a1f78a src/makemake.in
--- a/src/makemake.in	Thu Mar 12 13:23:02 2015 -0400
+++ b/src/makemake.in	Thu Mar 12 13:45:19 2015 -0400
@@ -1376,13 +1376,24 @@
   # Support for dynamic loading.
   eval "`./libtool --tag=CC --config | grep '^pic_flag='`"
   eval XCC_PICFLAG=\"${pic_flag}\"
-  eval "`./libtool --tag=CC --config | grep '^archive_cmds='`"
+  if [ ${HSYSOS} != cygwin ]; then
+    eval "`./libtool --tag=CC --config | grep '^archive_cmds='`"
+  else
+    archive_cmds='$CC -shared $libobjs $deplibs $compiler_flags ${absolute_sourcedir}/lisp.dll.a -o $output_objdir/$soname ${wl}--enable-auto-import ${wl}--out-implib=${dll}.a'
+  fi
   # http://article.gmane.org/gmane.comp.shells.bash.bugs:17158
   CC_set=${CC+set}
   CC_save=$CC
-  CC='${CC}' libobjs='$libs' deplibs='${CLFLAGS}' compiler_flags='${CFLAGS}' \
-    soname='$dll' lib='$lib' output_objdir='$dyndir' \
-    eval XCC_CREATESHARED=\"${archive_cmds}\"
+  if [ "${HSYSOS}" != cygwin ] ; then
+    CC='${CC}' libobjs='$libs' deplibs='${CLFLAGS}' compiler_flags='${CFLAGS}' \
+      soname='$dll' lib='$lib' output_objdir='$dyndir' \
+      eval XCC_CREATESHARED=\"${archive_cmds}\"
+  else
+    CC='${CC}' libobjs='$libs' deplibs='${CLFLAGS}' compiler_flags='${CFLAGS}' \
+      soname='$dll' lib='$lib' output_objdir='$dyndir' wl='-Wl,' \
+      absolute_sourcedir='$absolute_sourcedir' dll='$dll' \
+      eval XCC_CREATESHARED=\"${archive_cmds}\"
+  fi
   if test "$CC_set" = set
   then CC=$CC_save
   else unset CC
@@ -2604,7 +2615,7 @@
   echotab "\$(RANLIB) libnoreadline.a"
   echol
 fi
-if [ ${SHREXT} = .dll -a "${with_dynamic_modules}" != no ]; then
+if [ ${SHREXT} = .dll -a "${with_dynamic_modules}" != no -a "${HSYSOS}" != cygwin ]; then
   # win32 (SHREXT=.dll) requires all accessed libraries when linking a DLL.
   # i.e. when creating a dynamic module dll, we must give lisp.dll to the
   # linker, and when creating lisp.dll, we must give $(LIBS) to it.
